### DISCUSSION TITLE
Manual entityManager.registerSubscriber not working with transactional

### DISCUSSION THREAD

**Discussion by hungtcs** [mikro-orm/mikro-orm#3581] [Q&A] [answered]
This is my code:

```ts
@Injectable()
export class SomeService implements EventSubscriber<SomeEntity> {

  constructor(
      private readonly entityManager: EntityManager,) {
    this.entityManager.getEventManager().registerSubscriber(this);
  }

  public afterCreate(event: EventArgs<SomeEntity>) {
    console.log('------------ afterCreate ------------');
  }

  public async createSomeEntity() {
    return await this.entityManager.transactional(async (entityManager) => {
      const item = new SomeEntity({ name: '' });
      await entityManager.persistAndFlush(item);
      return item;
    });
  }

  public async createSomeEntity2() {
    const item = new SomeEntity({ name: '' });
    await this.entityManager.persistAndFlush(item);
    return item;
  }

}
```

When I call `createSomeEntity` the `afterCreate` not working, but when I call `createSomeEntity2`, the hook is working fine.

All the differences are only because I explicitly created a transaction.

Does anyone know why? Thanks!

**Comment by B4nan** [comment, marked as answer]
This does not make much sense to me, a subscriber is stateless, a singleton, you shouldnt work with it as with a regular service - you should not call its API from outside. 

What are you trying to do? Where are you calling this method from?

**Reply by hungtcs** [reply]
Hi @B4nan ,

I put subscriber and service together is in order to call the method of service in `afterCreate` hook.
Everything work well until I wrote a transactional.

**Reply by hungtcs** [reply]
I noticed that the two `entityManager` instances are different at this time.
Is the `this.entityManager.getEventManager().registerSubscriber(this)` only working with `this.entityManager`? not the entityManager inside transactional scope.

```ts
public async createSomeEntity() {
  return await this.entityManager.transactional(async (entityManager) => {
    console.log(this.entityManager === entityManager); // <-- false
    const item = new SomeEntity({ name: '' });
    await entityManager.persistAndFlush(item);
    return item;
  });
}
```

**Reply by B4nan** [reply]
> I put subscriber and service together is in order to call the method of service in afterCreate hook.

I just told you thats not right. And I am asking what are you trying to do, not _how_, but _what_. 

> I noticed that the two entityManager instances are different at this time.

Yes, that is how `em.transactional` works, you have a different fork inside the callback.

> Is the this.entityManager.getEventManager().registerSubscriber(this) only working with this.entityManager? not the entityManager inside transactional scope.

Forks made via `em.transactional` will have their own `EventManager` instance, so yes, those calls are not the same. If you register subscriber on the fork, it will be registered only there, not in upper scope.

**Reply by hungtcs** [reply]
Thank you for your explanation, I just realized I should change my mind, now I have solved the problem in another way.
